libs/common/src/blockchain/blockchain.service.ts

```diff
@@ -157,4 +156,71 @@ export class BlockchainService implements OnApplicationBootstrap, OnApplicationS
+    const txReceiptPromises: Promise<{ success: boolean; blockHash?: BlockHash; capacityWithDrawn?: string; error?: RegistryError }>[] = blockList.map(async (blockNumber) => {
+      const blockHash = await this.getBlockHash(blockNumber);
+      const block = await this.getBlock(blockHash);
+      const txInfo = block.block.extrinsics.find((extrinsic) => extrinsic.hash.toString() === txHash.toString());
+
+      if (!txInfo) {
+        return { success: false };
+      }
+
+      this.logger.verbose(`Found tx ${txHash} in block ${blockNumber}`);
+      const at = await this.api.at(blockHash.toHex());
+      const eventsPromise = firstValueFrom(at.query.system.events());
+
+      let isTxSuccess = false;
+      let totalBlockCapacity: bigint = 0n;
+      let txError: RegistryError | undefined;
+
+      try {
+        const events = await eventsPromise;
+
+        events.forEach((record) => {
+          const { event } = record;
+          const eventName = event.section;
+          const { method } = event;
+          const { data } = event;
+          this.logger.debug(`Received event: ${eventName} ${method} ${data}`);
+
+          // find capacity withdrawn event
+          if (eventName.search('capacity') !== -1 && method.search('Withdrawn') !== -1) {
+            // allow lowercase constructor for eslint
+            // eslint-disable-next-line new-cap
+            const currentCapacity: u128 = new u128(this.api.registry, data[1]);
+            totalBlockCapacity += currentCapacity.toBigInt();
+          }
+
+          // check custom success events
+          if (successEvents.find((successEvent) => successEvent.pallet === eventName && successEvent.event === method)) {
+            this.logger.debug(`Found success event ${eventName} ${method}`);
+            isTxSuccess = true;
+          }
+
+          // check for system extrinsic failure
+          if (eventName.search('system') !== -1 && method.search('ExtrinsicFailed') !== -1) {
+            const dispatchError = data[0] as DispatchError;
+            const moduleThatErrored = dispatchError.asModule;
+            const moduleError = dispatchError.registry.findMetaError(moduleThatErrored);
+            txError = moduleError;
+            this.logger.error(`Extrinsic failed with error: ${JSON.stringify(moduleError)}`);
+          }
+        });
+      } catch (error) {
+        this.logger.error(error);
+      }
+      this.logger.debug(`Total capacity withdrawn in block: ${totalBlockCapacity.toString()}`);
+      return { success: isTxSuccess, blockHash, capacityWithDrawn: totalBlockCapacity.toString(), error: txError };
+    });
+    const results = await Promise.all(txReceiptPromises);
+    const result = results.find((receipt) => receipt.blockHash !== undefined);
+    this.logger.debug(`Found tx receipt: ${JSON.stringify(result)}`);
+    return result ?? { success: false };
+  }
```

Comment:
A generic block crawler that takes events to search and returns details of tx, i.e. success or failure, what error is encounter and in which block hash it was found